### PR TITLE
Fybrik Application taxonomy - fixing Reconciler validation 

### LINKED ISSUE #537: Fybrik Application Taxonomy
Task list to create and use a Fybrik Application Taxonomy:
- [x] Finalize the values in Fybrik application JSON schema taxonomy
- [x] Validate schema using @roee88 's [schema validator](https://github.com/roee88/the-mesh-for-data/tree/taxonomy-tool/pkg/taxonomy) #533 
- [x] Move all taxonomy json files to `charts/files` to create config maps #652 
- [x] Write JSON schema taxonomy to config maps #652 
- [x]  Mount config maps to manager deployment as volumeMounts #652 
- [x] Remove smart comments of kubebuilder with hard-coded enums #645 
- [x] Read taxonomy from config map in Fybrik application webhook and validate Fybrik application #645 
- [x] [Check](https://github.com/mesh-for-data/mesh-for-data/blob/664cdbe977f082315194b1a6472661a7036aa1ed/manager/README.md#run-and-debug-locally) if taxonomy validation is working in webhook
- [x] Add validation code to Reconciler and execute if webhooks flag is disabled
- [x] Remove data API strings from the manager code @shlomitk1

### PR DESCRIPTION
**Pull request by eletonia**
This PR makes the following changes:

- [x] Adds 2 new fields to `FybrikApplication` Status: `ValidApplication` and `ValidatedGeneration`
- [x] Reconciler checks the `ValidatedGeneration` field to see if the application has already been validated
- [x] Reconciler sets `ValidApplication` to `true` or `false` depending on if the `FybrikApplication` passed the taxonomy check
- [x] Reconciler shows specific validation errors in the manager logs and `FybrikApplication` status
- [x] Modifies controller unit tests to update generation version
- [x] Checks `ENABLE_WEBHOOKS` flag to see if validation is needed in Reconciler

Fixes #537

### CODE REVIEW

**Review by ronenkat — COMMENTED** [review]
Do we need two status indicators?
why not just `validatedGeneration` ? with the semantics of nil for never verified, a generation as the last one validated.
if current generation == `validatedGeneration` then we are good.
What is the down size for this approach? 
@shlomitk1 do you have a point of view?

**Comment by shlomitk1** [comment]
> 
> 
> Do we need two status indicators?
> why not just `validatedGeneration` ? with the semantics of nil for never verified, a generation as the last one validated.
> if current generation == `validatedGeneration` then we are good.
> What is the down size for this approach?
> @shlomitk1 do you have a point of view?

The down size is that for an invalid application a new reconcile (for example, after a status update) will try to validate it again.  If this is the app without a plotter, the second reconcile will be the last one. If there is a plotter from a previous version, a reconcile will be called on each plotter update.

**Comment by ronenkat** [comment]
> > Do we need two status indicators?
> > why not just `validatedGeneration` ? with the semantics of nil for never verified, a generation as the last one validated.
> > if current generation == `validatedGeneration` then we are good.
> > What is the down size for this approach?
> > @shlomitk1 do you have a point of view?
> 
> The down size is that for an invalid application a new reconcile (for example, after a status update) will try to validate it again. If this is the app without a plotter, the second reconcile will be the last one. If there is a plotter from a previous version, a reconcile will be called on each plotter update.

@shlomitk1, what you are saying is that any status update (triggered by a plotter) will advance the appVersion, triggering a revalidation --- this is the current code behaviour in the PR --- the validation is triggered by observedStatus.ValidatedGeneration != appVersion.

From looking at the PR the boolean indication is just for visibility to a user, and is not used as a control logic. @eletonia am I missing something in the logic?

**Comment by eletonia** [comment]
> > > Do we need two status indicators?
> > > why not just `validatedGeneration` ? with the semantics of nil for never verified, a generation as the last one validated.
> > > if current generation == `validatedGeneration` then we are good.
> > > What is the down size for this approach?
> > > @shlomitk1 do you have a point of view?
> > 
> > 
> > The down size is that for an invalid application a new reconcile (for example, after a status update) will try to validate it again. If this is the app without a plotter, the second reconcile will be the last one. If there is a plotter from a previous version, a reconcile will be called on each plotter update.
> 
> @shlomitk1, what you are saying is that any status update (triggered by a plotter) will advance the appVersion, triggering a revalidation --- this is the current code behaviour in the PR --- the validation is triggered by `observedStatus.ValidatedGeneration != appVersion`.
> 
> From looking at the PR the boolean indication is just for visibility to a user, and is not used as a control logic. @eletonia am I missing something in the logic?

The latest commit changes the `ValidApplication` boolean to an enum with the options of `Unknown`, `True`, and `False`. This is now being used in the logic of the Reconciler as suggested by @shlomitk1

**Comment by shlomitk1** [comment]
> @shlomitk1, what you are saying is that any status update (triggered by a plotter) will advance the appVersion, triggering a revalidation --- this is the current code behaviour in the PR --- the validation is triggered by `observedStatus.ValidatedGeneration != appVersion`.
> 
> From looking at the PR the boolean indication is just for visibility to a user, and is not used as a control logic. @eletonia am I missing something in the logic?

@ronenkat No, appVersion is not changed. Perhaps I have misunderstood your suggestion for ValidatedGeneration. If the suggestion is to update this field only for valid applications, then we'll get multiple validations for invalid applications as I mentioned. Otherwise, if every validation updates the field regardless of the result, then the next reconcile will not validate again but will not have enough information to know that the reconcile must not proceed. In other words, we need to know 1) when the validation was done to avoid extra validations and 2) validation result to avoid working with an invalid application.

Regarding the usage: if the application is invalid (regardless whether the validation was made in this reconcile or earlier, for the same generation), reconcile stops immediately.

File: manager/apis/app/v1alpha1/fybrikapplication_types.go [hunk 1]
```diff
@@ -189,6 +197,14 @@ type FybrikApplicationStatus struct {
 	// +optional
 	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
 
+	// ValidatedGeneration is the version of the FyrbikApplication that has been validated with the taxonomy defined.
```
Comment (shlomitk1) [inline manager/apis/app/v1alpha1/fybrikapplication_types.go:200]: @eletonia Somehow after the merge the new structures appear twice.

Comment (eletonia) [reply manager/apis/app/v1alpha1/fybrikapplication_types.go:192]: I just removed the duplicate variables in the latest commit

File: manager/controllers/app/fybrikapplication_controller.go [hunk 2]
```diff
@@ -72,18 +73,27 @@ func (r *FybrikApplicationReconciler) Reconcile(ctx context.Context, req ctrl.Re
 	observedStatus := applicationContext.Status.DeepCopy()
 	appVersion := applicationContext.GetGeneration()
 
-	// check if observed generation is not the same as appVersion
-	if observedStatus.ObservedGeneration != appVersion {
+	// check if webhooks are enabled and application has been validated before or if validated application is outdated
+	if os.Getenv("ENABLE_WEBHOOKS") == "false" && (string(applicationContext.Status.ValidApplication) == "" || observedStatus.ValidatedGeneration != appVersion) {
```
Comment (shlomitk1) [inline manager/controllers/app/fybrikapplication_controller.go:77]: Better make it` != "true"` to capture undefined variables as well

**Review by shlomitk1 — APPROVED** [review]
A minor comment regarding checking the environment variable.

File: manager/controllers/app/fybrikapplication_controller.go (hunk 2 above)
Comment (eletonia) [reply manager/controllers/app/fybrikapplication_controller.go]: Updated in the latest commit

**Review by ronenkat — APPROVED** [review]

**Pull request merged**